File: NoofIslands-BFS.py
#Time Complexity: O(m*n)
#space Complexity: O(min(m,n))
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0
        
        m = len(grid)
        n = len(grid[0])
        dirs = [[0,1],[1,0],[-1,0],[0,-1]]
        
        cnt = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    cnt += 1
                    q = deque()
                    q.append((i,j))
                    grid[i][j] = "0"
                    
                    while q:
                        curr = q.popleft()
                        for d in dirs:
                            nr = curr[0] + d[0]
                            nc = curr[1] + d[1]
                            
                            if nr >= 0 and nc >= 0 and nr < m and nc < n:
                                if grid[nr][nc] == "1":
                                    q.append((nr, nc))
                                    grid[nr][nc] = "0"
        return cnt
                    
        

File: NoofIsLandsBFS.java
class Solution {
    //tc = o(m*n)
    //sc=o(m*n)
    public int numIslands(char[][] grid) {
        if(grid == null || grid.length ==0) return 0;
        
        int m = grid.length;
        int n = grid[0].length;
        int[][] dirs = new int[][]{{0,-1},{0,1},{-1,0},{1,0}};
        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();
        int count =0;
        
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j] == '1')
                {
                    q1.add(i);
                    q2.add(j);
                    grid[i][j] = '0';
                    count++;
                    
                    //bfs
                    while(!q1.isEmpty())
                    {
                       int cr = q1.poll();
                       int cc = q2.poll();
                        
                        for(int[] dir : dirs)
                        {
                            int nr = dir[0] + cr;
                            int nc = dir[1] + cc;
                            //boundary conditions
                            if(nr >= 0 && nc >=0 && nr <m && nc < n && grid[nr][nc] == '1')
                            {
                                q1.add(nr);
                                q2.add(nc);
                                
                                grid[nr][nc] = '0';
                                
                            }
                            
                        }
                    }
                    
                }
            }
        }
        
        return count;
        
    }
}